community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/checkpoint/CheckPointScheduler.java

```diff
@@ -19,22 +19,32 @@
  */
 package org.neo4j.kernel.impl.transaction.log.checkpoint;
 
-import java.io.IOException;
+import java.util.Arrays;
 import java.util.function.BooleanSupplier;
 
 import org.neo4j.function.Predicates;
 import org.neo4j.kernel.impl.store.UnderlyingStorageException;
 import org.neo4j.kernel.impl.util.JobScheduler;
+import org.neo4j.kernel.internal.DatabaseHealth;
 import org.neo4j.kernel.lifecycle.LifecycleAdapter;
 
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static org.neo4j.kernel.impl.util.JobScheduler.Groups.checkPoint;
 
 public class CheckPointScheduler extends LifecycleAdapter
 {
+    /**
+     * The max number of consecutive check point failures that can be tolerated before treating
+     * check point failures more seriously, with a panic.
+     */
+    static final int MAX_CONSECUTIVE_FAILURES_TOLERANCE = 3;
+
     private final CheckPointer checkPointer;
     private final JobScheduler scheduler;
     private final long recurringPeriodMillis;
+    private final DatabaseHealth health;
+    private volatile Throwable[] failures = new Throwable[MAX_CONSECUTIVE_FAILURES_TOLERANCE];
```

Comment:
If your reason for using `volatile` is safe publication, `final` would take care of that just as well (as long as you don't publish the `CheckPointScheduler` to other threads before the constructor has completed).

I don't think using `volatile` as a _better-safe-than-sorry_ directive is helpful, since it does not only mean something to the runtime, it also means something to the developers that will maintain this code, and then a `volatile` field that is never updated is really confusing. The fact that the type of the field is an array makes it even more confusing, especially to someone who doesn't know Java really well, since it might make you think that it is an array of volatile references to `Throwable` objects, it is not, it is a volatile reference to a single array of normal references to `Throwable` objects. The `volatile` does nothing (or at least very little) to help you ensure that the actual references in the array are safely published.